[ENHANCEMENT] Defer resource initialization, add templating, unit tes…

jeffkpayne opened the pull request:
…ts, clean up

Addresses #18 with the following specific changes:

* Defer initialization of DataContext and other resources
* Mark additional vars templated in BigQuery operator
* Add type hints and update docstrings
* Add unit and integration test coverage
* Add example DAG setup conveniences
* Add explicit Airflow version handling
* Update README with virtual env and testing process

jeffkpayne commented:
@cla-bot check

talagluck commented on README.md:
```diff
@@ -42,24 +42,96 @@ from great_expectations_provider.operators.great_expectations_bigquery import Gr
 2. Make the `great_expectations` and `data` directories in `include/` available in your environment.
 3. Change the `data_file` and `ge_root_dir` paths in your DAG file to point to the appropriate places.
-4. Change the paths in `great-expectations/checkpoints/*.yml` to point to the absolute path of your data files. 
+4. Change the paths in `great-expectations/checkpoints/*.yml` to point to the absolute path of your data files.
+5. Change the value of [`enable_xcom_pickling`](https://github.com/apache/airflow/blob/master/airflow/config_templates/default_airflow.cfg#L181) to `true` in your airflow.cfg
```
Given that the instructions for Astronomer require copying stuff over to an Astronomer project directory, what would you think about having another directory that we could point to for a vanilla Airflow run (with the directories wired up to work)?

jeffkpayne replied:
Ah, yeah, that makes sense.  When you say "directory", you mean another `great-expectations` directory called, for instance, `great-expectations-vanilla-airflow`, with the appropriate absolute paths for a typical vanilla AF installation, I assume?

talagluck replied:
Exactly. Maybe have a structure sort of like this

```
airflow-provider-great-expectations
 -> examples
   -> astronomer
     -> include
       -> ...  
   -> vanilla_airflow
     -> ... 
```

talagluck commented on include/great_expectations/expectations/.ge_store_backend_id:
```diff
@@ -0,0 +1 @@
+store_backend_id = 28fc8bae-d91c-4ca3-8c9f-b0270ea34e7d
```
I need to double-check with the team if this is the best way to set a store_backend_id for a case like this (it might mess with our usage stats).

jeffkpayne replied:
Yeah, I honestly wasn't very clear on what that was used for and didn't notice that it was included in the commits.  If it's for collecting usage stats, then it should be ignored and not committed, I would guess.  Let me know.

jeffkpayne replied:
Removed the usage stats config section completely and ignored the `.ge_store_backend_id` file.

talagluck commented on requirements.txt:
```diff
@@ -1 +1,5 @@
-great-expectations>=0.13.14   # package
+--index-url https://pypi.python.org/simple/
+# Install the target package and its dependencies
+-e .
```
I haven't seen `-e .` in a `requirements.txt` file before. Is there a reason why we would want to use this here?

jeffkpayne replied:
It defers the dependencies to `setup.py` and seemed to be an accepted and logical approach for reducing the duplication of dependencies between the two files, especially for the purposes of testing during development.  I'll reference https://caremad.io/posts/2013/07/setup-vs-requirement/ for a better explanation.  I'm in no way tied to this approach :)

talagluck replied:
Thanks for the explanation!

talagluck commented on tests/__init__.py:
```diff
@@ -0,0 +1,9 @@
+# Solution recommended by
+# http://python-guide-pt-br.readthedocs.io/en/latest/writing/structure/
+# import os
+# import sys
+# import airflow.utils.db
+#
+# sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../../")))
+
+# airflow.utils.db.initdb()
```
What is this for? The link does not appear to work.

jeffkpayne replied:
Gah, the commented out code should have been deleted.

talagluck commented on tests/operators/test_great_expectations.py:
```diff
@@ -0,0 +1,149 @@
+from great_expectations.exceptions.exceptions import GreatExpectationsError
+
+
+log = logging.getLogger(__name__)
```
Changing to be more consistent with the `great_expectations` codebase.
```suggestion
logger = logging.getLogger(__name__)
```

jeffkpayne replied:
See my other response around the use of `self.log` instead of the global logger instance...

jeffkpayne replied:
Gah, never mind.  Just noticed that this is the test.  LOL.  Will change from `log` to `logger`.

talagluck commented on tests/operators/test_great_expectations.py:
```diff
@@ -0,0 +1,149 @@
+        self.assertRaises(GreatExpectationsError, operator.execute, context={})
+
+    def test_great_expectations_operator__valid_checkpoint(self):
+        # This is ugly, but the older version of checkpoint configs that this supports
+        # doesn't seem to allow relative paths for the data assets being validated.
```
If you look at the main `great_expectations` repo, you'll see similar, but maybe more realized patterns that might be helpful, for instance in `tests/data_context/test_data_context.py::test_render_full_static_site_from_empty_project` and in `tests/integration/test_script_runner.py::_execute_integration_test`

jeffkpayne replied:
I'd, personally, prefer to use `pytest` along with all the features and fixtures it provides (like `tmp_path`), but there was a desire by Plinio to support unit tests being run with a simple `python -m unittest ...` and to not directly rely on `pytest`.

If you'd prefer that we just assume `pytest` is being used to drive the tests, great.  Let me know and I can migrate the unit tests to `pytest` style tests.

talagluck replied:
That's great context to have. I hear Plinio's point, but we do have the eventual goal of incorporating this project within the main the Great Expectations codebase (though we do not have a definite timeline for this). At that point, we'll need to migrate to the standard `pytest` way. My inclination is to just do that now since it makes things easier for this type of maintenance in the short-term and the long-term, unless it makes things significantly more difficult for the Astronomer folks (and then it would be good to discuss).

jeffkpayne replied:
OK, so should we get Plinio's feedback now or just make the changes first?  I'm fine either way...

talagluck replied:
I think let's check with @pgzmnk to confirm first.

pgzmnk replied:
While we _encourage_ `unittest` as a standard across Airflow providers, using `pytest` makes sense given the reasoning above.

jeffkpayne replied:
Thanks, @pgzmnk.  I went ahead and moved to the `pytest` only tests.

CC @talagluck

talagluck replied:
Thanks, both!

talagluck commented on great_expectations_provider/operators/great_expectations.py:
```diff
@@ -17,97 +17,160 @@
 import great_expectations as ge
 from great_expectations.checkpoint import LegacyCheckpoint
-
-log = logging.getLogger(__name__)
```
Why remove the logger?

jeffkpayne replied:
The operator logging is done with `self.log.info()`, etc, as is the recommended approach for AF operators, AFAIK.  `self.log` is provided by the `LoggingMixin` super class.  For instance, line 161.

talagluck commented on great_expectations_provider/operators/great_expectations.py:
```diff
@@ -117,34 +180,198 @@ def execute(self, context):
+    )
+
+    @apply_defaults
+    def __init__(
```
Is there a reason why you removed the `email_to`, etc. parameters and functionality for the BigQuery operator? (I would push toward making it an optional parameter, but including it).

jeffkpayne replied:
That was a judgement call.  Three things pushed me to remove it:

- Callers can use the built-in email notification functionality of GE itself (actions, I believe)
- Callers can use the generic hook provided by the new `validation_failure_callback` parameter (allowing the caller to use whatever kind of notification they want)
- I thought it was strange that the email alert functionality only existed on the BQ operator (maybe I'm just not clear on *why* it only existed there)

I thought the resulting operators were a little more focused on GE specific functionality.  Of course, I'm fine adding it back or providing a `validation_failure_callback` implementation that handles sending an alert email.  Let me know.

talagluck replied:
Hi @jeffkpayne - some excellent points here. The history of this predates me a bit, but I think it was basically that someone had built and contributed the BigQuery-specific operator first, and when we saw that, someone on my team went back and built the more generic parent-class Operator. My hypothesis is that that person probably didn't want to reduce functionality that had been built by the original contributor, but also didn't want to add this specific functionality to the parent-class.

Given all your work here to clean things up and make them more consistent, I think you're right to remove this here. If we get some complaints, I'd be OK potentially reverting it down the line, but I think this is a great initial direction to go. That said, would you mind adding an additional note to the README along the lines of: "We have removed the email functionality from the top level for the sake of consistency in our Operators, but if you need to access it you can do so in <any of the ways you mentioned above>"?

jeffkpayne replied:
Ah, OK, it all makes more sense now :).  Yes, I will add a note to the README wrt the email func being removed.  Thanks!

jeffkpayne replied:
Hm, the more I think about it, and given the above context, it seems like the BQ-specific operator should really just be a "GCS-based data context" operator.  Either that or it's just removed completely and a hook be added to create any DataContext, local file based like in `GreatExpectationsOperator` or GCS based like in `GreatExpectationsBigQueryOperator`, at run-time.  These types of changes seem like a future version to me, and maybe one that is even opinionated about which GE API version is used (v3), but wanted to bring it up now...

Anyway, I'll update the README for now ;)

talagluck replied:
@jeffkpayne - I think that your intuition here is spot-on. We're due for some additional overhauls of this project (the migration to V3 is a big one), and so that I think this refactor would be a great candidate for one of these big changes. I don't know when I'll have the resources to devote a designated chunk of work to that, but let me know if you're interested, and I'm happy to spend some time in closer collaboration if that's helpful.

jeffkpayne replied:
@talagluck I'd be up for helping out with that.  Your assistance would be great, given that I'm not fully up to speed on v2 vs v3 diffs.  I'd suggest that be a future effort, outside of this PR.  Is that what you had in mind?

pgzmnk replied:
We too can dedicate developer resources within our area of competence to this matter for a future PR.

talagluck replied:
Thanks, @jeffkpayne and @pgzmnk! Yes this would be a future effort in a new PR. Let's hop on a call sometime in the next couple of weeks to discuss if that works for you.

talagluck reviewed (commented):
This is awesome - thanks so much for this contribution, @jeffkpayne! I love how thorough the README is, and appreciate all of the improvements to organization, documentation, and testing.

I have a few questions/suggestions, and then one question that I need to bring to the team on our side.

I've been able to test this with vanilla Airflow, but not yet with Astronomer, and my goal is to do that tomorrow.

jeffkpayne commented:
@talagluck I think I got to all the comments.  Thanks for the great feedback/questions.  I'll make the obvious changes ASAP and await your response to the other comments.

talagluck reviewed (commented):
Thanks for the insight and thorough explanations, @jeffkpayne! I responded to your comments below.

A couple of other notes:
- I believe that the .ge_store_backend_id will be necessary for the project to run properly (and by deleting it, it will just autogenerate a new one the next time things are run). Would you mind adding it back in, and instead, modify the backend id to be something that is obviously artificial (e.g. abcdabcd-1111-2222-3333-abcdabcdabcd), which will enable us to monitor it in the usage stats?
- Were you able to sign the CLA? I don't believe that CLA bot is enabled in this repo, but I was able to check manually, and I didn't see your username there. Please let me know if you have any questions or issues.

jeffkpayne commented:
@talagluck Will add the backend ID back in with a value of "abcdabcd-1111-2222-3333-abcdabcdabcd".

I signed the company level CLA for Bombora Inc.  Here's a screen shot of the successful github check:
<img width="972" alt="Screen Shot 2021-08-19 at 2 45 09 PM" src="https://user-images.githubusercontent.com/43057/130149406-cccddb3e-9fa3-49e7-9452-c3850d941ae9.png">

talagluck commented:
Sounds good! This is looking great. Let's just wait to a bit to hear from Plinio about the tests. Once that's resolved, and we add in the .ge_store_backend_id, we should be all set!

talagluck commented on include/great_expectations/great_expectations.yml:
```diff
@@ -124,6 +124,6 @@ data_docs_sites:
 anonymous_usage_statistics:
-  data_context_id: 28fc8bae-d91c-4ca3-8c9f-b0270ea34e7d
-  enabled: true
+  data_context_id: 00000000-0000-0000-0000-000000000000
+  enabled: false
```
Hi @jeffkpayne - could we change this `data_context_id` to the one that I added in the comment on `.ge_store_backend_id`? Or something else that looks artificial, but is also not all 0s (we use the all 0s for our testing). Would you mind also enabling usage stats? I talked to our analytics folks, and we figured it might be useful to see how often people were using this tutorial. If you have serious concerns about this, I'm open to leaving this as it is.

jeffkpayne replied:
Ah, yes! Lol. Sorry, I guess I picked the same testing value that you did. I'll take that as a compliment! Thanks. Doing that now.

jeffkpayne commented:
@talagluck I added the pytest versions of the unit tests in a `pytests` directory.  As soon as you and @pgzmnk agree on the unit tests framework, I'll just remove the obsolete tests ;)

jeffkpayne commented:
@talagluck Let me know and I'll squash commits.

talagluck commented on tests/operators/test_great_expectations.py:
```diff
@@ -0,0 +1,154 @@
+
+
+def test_great_expectations_operator__valid_checkpoint():
+    # This is ugly, but the older version of checkpoint configs that this supports
+    # doesn't seem to allow relative paths for the data assets being validated.
```
Hi @jeffkpayne ! Could we modify this comment a bit? Maybe something like
```suggestion
    # Using this approach since older version of checkpoint configs that this supports
    # don't allow relative paths for the data assets being validated.
```

talagluck replied:
(I think it's probably fine to remove the comment entirely - we do use a similar approach in the main Great Expectations codebase in situations like this)

talagluck commented:
Thanks for all the revisions, @jeffkpayne ! I have one more request to update a comment, and then we are all set to merge.

jeffkpayne commented:
@talagluck Awesome.  Made that last change, squashed commits, and updated the commit message to include all pertinent into 😉 

Thanks again!  Can't wait to see the package on PyPi!  😀

talagluck reviewed (approved):
LGTM! Thanks for all the hard-work and updates!

talagluck commented:
Thanks, @jeffkpayne ! Just merged, we'll cut a new release in the next couple of days.
